bundles/org.openhab.binding.thekeys/src/main/java/org/openhab/binding/thekeys/internal/smartlock/TheKeysSmartlockHandler.java

```diff
@@ -0,0 +1,216 @@
+        super(thing);
+    }
+
+    @Override
+    public void initialize() {
+        config = getConfigAs(TheKeysSmartlockConfiguration.class);
+        gateway = (TheKeysGatewayHandler) getBridge().getHandler();
+        lastLog = -1;
+        updateStatus(ThingStatus.UNKNOWN, ThingStatusDetail.NOT_YET_READY,
+                "@text/message.status.wait-data-from-gateway");
+
+        checkIfBridgeOffline();
+    }
+
+    @Override
+    public void handleCommand(ChannelUID channelUID, Command command) {
+        if (command instanceof RefreshType) {
+            scheduler.submit(this::fetchAndUpdateLockStatus);
+        } else if (CHANNEL_LOCK.equals(channelUID.getId())) {
+            scheduler.submit(() -> lock((OnOffType) command));
+        }
+    }
+
+    /**
+     * Lock command
+     */
+    private void lock(OnOffType open) {
+        try {
+            updateState(CHANNEL_SYNC_IN_PROGRESS, OnOffType.ON);
+            if (open == OnOffType.ON) {
+                gatewayApi().open(config.lockId);
+            } else {
+                gatewayApi().close(config.lockId);
+            }
+        } catch (Exception e) {
+            logger.debug("Fail to execute {} request", open == OnOffType.OFF ? "open" : "close", e);
+            updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.COMMUNICATION_ERROR, e.getMessage());
+        }
+    }
+
+    private void checkIfBridgeOffline() {
+        Bridge bridge = getBridge();
+        if (bridge == null) {
+            updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.BRIDGE_UNINITIALIZED);
+        } else if (bridge.getStatus() != ThingStatus.ONLINE) {
+            updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.BRIDGE_OFFLINE);
+        }
+    }
+
+    /**
+     * Fetch the lock state from the gateway
+     *
+     * @return true if succeeded
+     */
+    private boolean fetchAndUpdateLockStatus() {
+        try {
+            updateState(CHANNEL_SYNC_IN_PROGRESS, OnOffType.ON);
+            LockerStatusDTO lockStatus = gatewayApi().getLockStatus(config.lockId);
+            if ("ko".equals(lockStatus.getStatus())) {
+                throw new TheKeysError(
+                        "Request failed with code " + lockStatus.getCode() + ". " + lockStatus.getCause());
+            }
+
+            updateState(CHANNEL_STATUS, new StringType(lockStatus.getStatus()));
+            updateState(CHANNEL_LOCK, OnOffType.from(!lockStatus.isClosed()));
+            updateState(CHANNEL_BATTERY_LEVEL, new DecimalType(getBatteryLevel(lockStatus.getBattery())));
+            updateState(CHANNEL_LOW_BATTERY, OnOffType.from(isLowBattery(lockStatus.getBattery())));
+            updateState(CHANNEL_POSITION, new DecimalType(lockStatus.getPosition()));
+            updateState(CHANNEL_RSSI, new DecimalType(lockStatus.getRssi()));
+            updateState(CHANNEL_LAST_SYNC, new DateTimeType(ZonedDateTime.now()));
+
+            updateProperty(PROPERTY_VERSION, String.valueOf(lockStatus.getVersion()));
```

Comment:
Rather use `ThingType.PROPERTY_FIRMWARE_VERSION` in case this is a firmware version.